src/Ocelot.Provider.Polly/OcelotBuilderExtensions.cs

```diff
@@ -1,44 +1,61 @@
-        };
-        return AddPolly<PollyQoSProvider>(builder, GetDelegatingHandler, errorMapping);
-    }
-
-    private static DelegatingHandler GetDelegatingHandler(DownstreamRoute route, IHttpContextAccessor contextAccessor, IOcelotLoggerFactory loggerFactory)
-        => new PollyPoliciesDelegatingHandler(route, contextAccessor, loggerFactory);
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+using Ocelot.Configuration;
+using Ocelot.DependencyInjection;
+using Ocelot.Errors;
+using Ocelot.Errors.QoS;
+using Ocelot.Logging;
+using Ocelot.Provider.Polly.Interfaces;
+using Ocelot.Requester;
+
+using Polly.CircuitBreaker;
+using Polly.Timeout;
+
+namespace Ocelot.Provider.Polly;
+
+public static class OcelotBuilderExtensions
+{
+    public static readonly Dictionary<Type, Func<Exception, Error>> DefaultErrorMapping = new Dictionary<Type, Func<Exception, Error>>
+    {
+        {typeof(TaskCanceledException), CreateRequestTimedOutError},
+        {typeof(TimeoutRejectedException), CreateRequestTimedOutError},
+        {typeof(BrokenCircuitException), CreateRequestTimedOutError},
+        {typeof(BrokenCircuitException<HttpResponseMessage>), CreateRequestTimedOutError},
+    };
+
+    private static Error CreateRequestTimedOutError(Exception e) => new RequestTimedOutError(e);
+
+    public static IOcelotBuilder AddPolly<T>(this IOcelotBuilder builder,
+            QosDelegatingHandlerDelegate delegatingHandler,
+            Dictionary<Type, Func<Exception, Error>> errorMapping)
+            where T : class, IPollyQoSProvider<HttpResponseMessage>
+    {
+        builder.Services
+            .AddSingleton(errorMapping)
+            .AddSingleton<IPollyQoSProvider<HttpResponseMessage>, T>()
+            .AddSingleton(delegatingHandler);
+
+        return builder;
+    }
+    
+    public static IOcelotBuilder AddPolly<T>(this IOcelotBuilder builder, Dictionary<Type, Func<Exception, Error>> errorMapping)
+        where T : class, IPollyQoSProvider<HttpResponseMessage> =>
+        AddPolly<T>(builder, GetDelegatingHandler, errorMapping);
+
+    public static IOcelotBuilder AddPolly<T>(this IOcelotBuilder builder, QosDelegatingHandlerDelegate delegatingHandler)
+        where T : class, IPollyQoSProvider<HttpResponseMessage> =>
+        AddPolly<T>(builder, delegatingHandler, ErrorMapping);
+
+    public static IOcelotBuilder AddPolly<T>(this IOcelotBuilder builder)
+        where T : class, IPollyQoSProvider<HttpResponseMessage> =>
+        AddPolly<T>(builder, GetDelegatingHandler, ErrorMapping);
+
+    public static IOcelotBuilder AddPolly(this IOcelotBuilder builder)
+    {
+        return AddPolly<PollyQoSProvider>(builder, GetDelegatingHandler, ErrorMapping);
+    }
+
+    private static DelegatingHandler GetDelegatingHandler(DownstreamRoute route, IHttpContextAccessor contextAccessor, IOcelotLoggerFactory loggerFactory)

```

Comment:
One note!
This private method is reused in lines 44, 52, 56.
All referenced extensions are public. I'm not sure this default delegate should be private for public extensions...

Reply:
example of our custom implementation

 ```
// todo : to simplify with ocelot 23
    public static IOcelotBuilder AddTPPolly(this IOcelotBuilder builder)
    {
        var errorMapping = new Dictionary<Type, Func<Exception, Error>>
        {
            { typeof(TaskCanceledException),  static e => new RequestTimedOutError(e) },
            { typeof(TimeoutRejectedException), static  e => new RequestTimedOutError(e) },
            { typeof(BrokenCircuitException),  static e => new RequestTimedOutError(e) },
            { typeof(BrokenCircuitException<HttpResponseMessage>), static  e => new RequestTimedOutError(e) },
        };
        return builder.AddPolly<PollyQoSProvider>(GetDelegatingHandler, errorMapping);
    }

    private static DelegatingHandler GetDelegatingHandler(DownstreamRoute route, IHttpContextAccessor contextAccessor, IOcelotLoggerFactory loggerFactory)
        => new PollyResiliencePipelineDelegatingHandler(route, contextAccessor, loggerFactory);
```